### ISSUE TITLE
-Db_sanitize=address is freezing several parts of our build

### ISSUE THREAD

**Issue by Jehan** [mesonbuild/meson#11408]
**Describe the bug**

So I was building GIMP with `-Db_sanitize=address` and we had several pieces of the build which just kind of "locked". The `ninja` command would just stop at some commands and nothing would happen.

I tried to run the build with verbose output (`-v`), and with `-d explain` and a single job (`-j1`) but could not really figure out how to get any reliable output (i.e. that `ninja` was not really giving any meaningful info on why it just stopped). I have not tried the next step, i.e. checking meson code itself.

**To Reproduce**

- Configure [GIMP's `master` branch](https://gitlab.gnome.org/GNOME/gimp/) with `meson configure -Db_sanitize=address`
- Build the project, possibly one job at a time (to get a possibly simpler output), in verbose and explaining commands: `ninja -v -d explain -j1` (or just `ninja` but that's for trying to debug)

**Expected behavior**

The build goes to the end, just like the non-sanitized build.

**Actual behavior**

At some point, the build stops. Nothing happens, nothing more is outputted. Some blocks can sometimes be passed by `ctrl-c`-ing them, then re-running `ninja` (which is not a good sign either; could it mean the output is incomplete but the next ninja doesn't realize?).

Some blocks are just always blocking until I remove the rules from the meson files to force `ninja`  to go further. For instance, it would infinitely block on very basic [dummy rules](https://gitlab.gnome.org/GNOME/gimp/-/blob/18c21abf0c0954fd2d397a68bd95df963d1f0acc/build/windows/meson.build#L21-35) which simply runs `python3 -c 'exit()'`.

Then if I disable these, it blocks when running some [command in `vcs_tag()`](https://gitlab.gnome.org/GNOME/gimp/-/blob/18c21abf0c0954fd2d397a68bd95df963d1f0acc/meson.build#L1622-1628) or when running [`gnome.generate_gir()` or `gnome.generate_vapi()`](https://gitlab.gnome.org/GNOME/gimp/-/blob/18c21abf0c0954fd2d397a68bd95df963d1f0acc/libgimp/meson.build#L463).

Basically trying to build with `-Db_sanitize=address` is painful and I only get a partial build as I have to force-disable or trick the system in various pieces of the build.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? Native build
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) PureOS 10.0/Debian
* what Python version are you using e.g. 3.8.0 - Python 3.9.2
* what `meson --version` - 0.56.2
* what `ninja --version` if it's a Ninja build - 1.10.1

**Comment by tristan957** [comment]
All the flag is gonna do is add compiler flags to various command lines. I don't see how this could possibly be a Meson issue if I am being honest. This sounds pretty specific to your system or the GIMP build system if other people can corroborate the issues you are seeing on different systems. Running `ninja` under `gdb` would probably be useful.

**Comment by eli-schwartz** [comment]
> Some blocks are just always blocking until I remove the rules from the meson files to force `ninja` to go further. For instance, it would infinitely block on very basic [dummy rules](https://gitlab.gnome.org/GNOME/gimp/-/blob/18c21abf0c0954fd2d397a68bd95df963d1f0acc/build/windows/meson.build#L21-35) which simply runs `python3 -c 'exit()'`.

That is an extremely puzzling diagnostic and I'm not sure how -Db_sanitize=address could even have an effect on arbitrary custom_targets that run dummy python commands.

I assume you've also tried running `ninja build/windows/gimp-plug-ins.rc` verify that ninja is staging precisely this build edge?

**Comment by Jehan** [comment]
@tristan957 For the record, when building GIMP with our autotools build system, I do this:

```sh
export LDFLAGS="-fsanitize=address"
export CFLAGS="-fsanitize=address"
export CXXFLAGS="-fsanitize=address"
```

Then usual configure and build. And it works absolutely fine. I just had to deal with a few expected issues (disabling sanitizer on some build-only tools, and I disabled GObject-Introspection altogether as it doesn't [play well with sanitizers](https://gitlab.gnome.org/GNOME/gobject-introspection/-/issues/375)). In any case, GIMP builds, when there are issues, I get a broken build (as expected) with errors which can then be diagnose normally. In particular no weird freeze of any sort of the build, with no error whatsoever.
And we are talking about the exact same code using either the autotools or meson build system. Not a different codebase.

It's nice that meson has a flag so that we don't have to tweak environment variables, but I wonder what it does exactly to freeze the build this way.

@eli-schwartz 

> I assume you've also tried running ninja build/windows/gimp-plug-ins.rc verify that ninja is staging precisely this build edge?

No. I just relied on it being the last step showing in the output (with `-j1` to disable multi-jobs which could make confusing output). 

Anyway I'll re-do a meson build in a separate folder to test a bit further.

**Comment by Jehan** [comment]
Hmmm… ok so I installed a more recent meson (0.63.0, self-built; I could not install more recent versions as I failed to even compile GLib with these, but I'll try later and will report issues if needed) and now the sanitized build does not freeze anymore.

I assume that whatever happened with the older meson 0.56.2 was fixed since then. So closing!
